src/libtcod/renderer_gl.c

```diff
@@ -64,31 +222,49 @@ static int upload_gl_tile(struct TCOD_TilesetAtlasOpenGL* atlas, int tile_id) {
       dest.h,
       GL_RGBA,
       GL_UNSIGNED_BYTE,
-      atlas->tileset->pixels + (tile_id * atlas->tileset->tile_length));
+      tile_origin);
   glBindTexture(GL_TEXTURE_2D, 0);
   return 0;
 }
 /**
  *  Setup a atlas texture and upload the tileset graphics.
+ *
+ *  This will be a square texture whose dimensions are a power of 2, of
+ *   sufficient size to hold all of the tiles.
  */
 TCOD_NODISCARD
 static int prepare_gl_atlas(struct TCOD_TilesetAtlasOpenGL* atlas) {
+#ifndef TILE_PADDING
+  const int TILE_PADDING = 0;
+#endif
   GLenum gl_error = glGetError();
   if (gl_error) return TCOD_set_errorvf("Unexpected OpenGL error before texture allocation: %u", gl_error);
+  // prospective texture size; start at a minimum of 64x64?
   int new_size = atlas->texture_size ? atlas->texture_size : 64;
+  // get maximum texture size supported by GPU
   int max_size;
   glGetIntegerv(GL_MAX_TEXTURE_SIZE, &max_size);
+  // starting values for tiles per row and per column in the atlas texture
+  // end values may differ because tiles aren't necessarily square in dimension
   int columns = 1;  // Must be more than zero.
   int rows = 1;
+  // calculate a texture size large enough to hold the tileset
   while (1) {
     if (atlas->tileset->tile_width == 0 || atlas->tileset->tile_height == 0) {
       break;  // Avoid division by zero.
     }
-    columns = new_size / atlas->tileset->tile_width;
-    rows = new_size / atlas->tileset->tile_height;
+    // calculate how many rows and colums of tiles can fit in a square texture
+    //  whose dimensions are both new_size
+    // padding around tiles is accounted for by adding to the individual tile
+    //  dimension divisors, while padding around the remaining horizontal and
+    //  vertical edge is accounted for by subtracting from the prospective
+    //  texture size
+    columns = (new_size - TILE_PADDING) / (atlas->tileset->tile_width + TILE_PADDING);
+    rows = (new_size - TILE_PADDING) / (atlas->tileset->tile_height + TILE_PADDING);
```

Comment:
This can't be something simpler?  Like `columns = new_size / (atlas->tileset->tile_width + TILE_PADDING * 2);`?

I assume that the total tile size is the tile size plus the padding.

Reply:
> `clang-tidy` should be run on this code before you commit it.
> 
> Don't comment out code on version control, it should be deleted or stashed.

I warned you this wasn't ready for primetime!



> This can't be something simpler? Like `columns = new_size / (atlas->tileset->tile_width + TILE_PADDING * 2);`?
> 
> I assume that the total tile size is the tile size plus the padding.

No. There needs to be padding along one side of each dimension of the atlas, plus per-tile padding along the opposite sides of each dimension of each tile.

The total atlas size in a given dimension is therefore padding_size + numTiles.dimension * (tile_size.dimension + padding_size)

Reply:
I assume that if the tiles are laid out like this:
```
111222
111222
111222
```
A pixel of padding will be applied like this:
```
##########
#111##222#
#111##222#
#111##222#
##########
```

Reply:
But I might be wrong with the exact math.  I often program iteratively.

Reply:
PADDING=2 with my approach, used with 2x2 tiles:
```
##########
##########
##11##22##
##11##22##
##########
##########
```

Your suggestion would be less compact:
```
############
############
##11####22##
##11####22##
############
############
```

Your approach would allow for odd numbered padding values though, which I didn't consider. Let me know if you think this is worth pursuing.

Reply:
I'd like to go with my approach and then hardcode a padding value of 1.

Alternatively I'd like to figure out framebuffers so that I can skip padding entirely, but I can try that on my PR.

Reply:
Fixed. Also verified that (at least with `texture2DLod`) `TILE_PADDING` value of 1 works for both `GL_NEAREST` and `GL_LINEAR`, and that commenting out the macro still works.